src/plugins/ui_feature/geo_features.ts

```diff
@@ -0,0 +1,146 @@
+import * as GeoJSON from "geojson";
+import { ModelConfig } from "../../model";
+import { GeoChartRegion } from "../../types/chart/geo_chart";
+import { UIPlugin, UIPluginConfig } from "../ui_plugin";
+
+class LoadingError {}
+const LOADING = new LoadingError();
+export class GeoFeaturePlugin extends UIPlugin {
+  static getters = [
+    "getGeoJsonFeatures",
+    "geoFeatureNameToId",
+    "getGeoChartAvailableRegions",
+  ] as const;
+
+  private readonly geoJsonService: ModelConfig["external"]["geoJsonService"];
+
+  private geoJsonCache: { [region: string]: GeoJSON.Feature[] | null | LoadingError } = {};
+  private featureNameToIdCache: Record<string, Record<string, string | null | LoadingError>> = {};
+  private availableRegions: GeoChartRegion[] | undefined | LoadingError = undefined;
+
+  private fetchingPromise?: Promise<void>;
+  private promises: Promise<void>[] = [];
+
+  constructor(config: UIPluginConfig) {
+    super(config);
+    this.geoJsonService = config.external.geoJsonService;
+  }
+
+  getGeoChartAvailableRegions(): GeoChartRegion[] {
+    if (!this.geoJsonService) {
+      console.error("No geoJsonService provided to the model");
+      return [];
+    }
+
+    if (this.availableRegions instanceof LoadingError) {
+      return [];
+    }
+    if (this.availableRegions !== undefined) {
+      return this.availableRegions;
+    }
+
+    this.availableRegions = LOADING;
+    this.promises.push(
+      new Promise<void>(async (resolve) => {
+        this.availableRegions = (await this.geoJsonService?.getAvailableRegions()) || [];
+        resolve();
+      })
+    );
+
+    this.triggerFetching();
+    return [];
+  }
+
+  getGeoJsonFeatures(region: string): GeoJSON.Feature[] | undefined {
+    if (!this.geoJsonService) {
+      console.error("No geoJsonService provided to the model");
+      return;
+    }
+
+    const cachedGeoJson = this.geoJsonCache[region];
+
+    if (cachedGeoJson instanceof LoadingError) {
+      return undefined;
+    }
+    if (cachedGeoJson !== undefined) {
+      return cachedGeoJson ?? undefined;
+    }
+
+    this.geoJsonCache[region] = LOADING;
+    this.promises.push(
+      new Promise<void>(async (resolve) => {
+        const json = await this.geoJsonService?.getTopoJson(region);
+        this.geoJsonCache[region] = this.convertToGeoJson(json);
+        resolve();
+      })
+    );
+    this.triggerFetching();
+    return undefined;
+  }
+
+  geoFeatureNameToId(region: string, featureName: string): string | undefined {
```

Comment:
I think there's a way to simplify things here.
We could load everything (`orm.searchRead("res.country", ...)`, etc.) when calling `getTopoJson` (maybe renamed then).
Then `geoFeatureNameToId` could synchronous (and not async). At this point, I think we could get rid of `triggerFetching` if it's only for `getTopoJson` as it's probably called most of the time only once per spreadsheet and at most a few times if multiple regions are in the spreadsheet)